clients/venice-admin-tool/src/main/java/com/linkedin/venice/RecoverStoreMetadata.java

```diff
@@ -0,0 +1,333 @@
+package com.linkedin.venice;
+
+import com.linkedin.venice.controller.kafka.AdminTopicUtils;
+import com.linkedin.venice.controller.kafka.protocol.admin.AdminOperation;
+import com.linkedin.venice.controller.kafka.protocol.admin.StoreCreation;
+import com.linkedin.venice.controller.kafka.protocol.admin.ValueSchemaCreation;
+import com.linkedin.venice.controller.kafka.protocol.enums.AdminMessageType;
+import com.linkedin.venice.controller.kafka.protocol.serializer.AdminOperationSerializer;
+import com.linkedin.venice.controllerapi.ControllerClient;
+import com.linkedin.venice.controllerapi.ControllerClientFactory;
+import com.linkedin.venice.controllerapi.ControllerResponse;
+import com.linkedin.venice.controllerapi.D2ServiceDiscoveryResponse;
+import com.linkedin.venice.controllerapi.NewStoreResponse;
+import com.linkedin.venice.controllerapi.SchemaResponse;
+import com.linkedin.venice.controllerapi.UpdateStoreQueryParams;
+import com.linkedin.venice.exceptions.VeniceException;
+import com.linkedin.venice.helix.HelixAdapterSerializer;
+import com.linkedin.venice.helix.HelixStoreGraveyard;
+import com.linkedin.venice.helix.StoreJSONSerializer;
+import com.linkedin.venice.kafka.protocol.KafkaMessageEnvelope;
+import com.linkedin.venice.kafka.protocol.Put;
+import com.linkedin.venice.kafka.protocol.enums.MessageType;
+import com.linkedin.venice.message.KafkaKey;
+import com.linkedin.venice.meta.Store;
+import com.linkedin.venice.pubsub.PubSubTopicPartitionImpl;
+import com.linkedin.venice.pubsub.PubSubTopicRepository;
+import com.linkedin.venice.pubsub.api.PubSubConsumerAdapter;
+import com.linkedin.venice.pubsub.api.PubSubMessage;
+import com.linkedin.venice.pubsub.api.PubSubTopicPartition;
+import com.linkedin.venice.security.SSLFactory;
+import com.linkedin.venice.utils.Utils;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.TreeMap;
+import org.apache.helix.zookeeper.impl.client.ZkClient;
+
+
+public class RecoverStoreMetadata {
+  public static void recover(
+      ZkClient zkClient,
+      PubSubConsumerAdapter consumer,
+      Optional<SSLFactory> sslFactory,
+      String url,
+      String storeName,
+      boolean skipLastStoreCreation,
+      boolean doRepair,
+      List<String> graveyardClusterList,
+      String recoverCluster) throws IOException {
+
+    /**
+     * Try to see whether the store exists or not.
+     */
+    ControllerClient controllerClient =
+        ControllerClientFactory.getControllerClient(graveyardClusterList.get(0), url, sslFactory);
```

Comment:
I don't understand why this is a list? Is it in case the operator doesn't know which cluster the store was originally part of (in which case they could supply a list of all the clusters?)... here there seems to be an assumption that the store must be found in the graveyard of the first cluster, which seems to defeat the purpose of passing a list? Can you comment on why a list is useful and what inputs are valid?

Reply:
This is used to create a dummy ControllerClient to call `discoverCluster`, essentially, any cluster will work here.
I can try to make it easier to understand.

The reason to introduce a list of clusters since we don't know which clusters the deleted store belongs to.

Reply:
I am going to use this static function instead to mitigate the confusion:
```
    D2ServiceDiscoveryResponse discoveryResponse = ControllerClient.discoverCluster(url, storeName, sslFactory, 3);
    if (discoveryResponse.getCluster() != null) {
      throw new VeniceException(
          "Store: " + storeName + " exists in cluster: " + discoveryResponse.getCluster() + ", no need to recover");
    }
```